### PR TITLE
Add interactive tutorials to the documentation

### PR DESCRIPTION
**Pull request by mplattner**
#### Description

Add interactive tutorials to the documentation using pre-recorded ascii casts. The recording is fully automated.

A preview is available at https://plattner.me/mp-docs/ under `mitmproxy Tutorial`

This PR supersedes PR #4127 

#### Todo

- [x] Merge classes `CliDirector` and `MitmCliDirector`
- [x] Restructure ToC of docs (esp. `Getting Started` and `Features`)
- [x] Polishing
- [ ] ~Add `Proxy Configuration` step (?)~

#### Checklist

 - [x] I have updated tests where applicable.
 - [x] I have added an entry to the CHANGELOG.

### CODE REVIEW

**Review by mhils — COMMENTED** [review]
Very awesome! 🎉

**Comment by mplattner** [comment]
Very helpful comments, thank you. More details tomorrow 😀

File: docs/src/assets/style.scss [hunk 1]
```diff
@@ -92,3 +95,30 @@ h1, h2, h3, h4, h5, h6 {
 figure.has-border img {
     box-shadow: 0 0 20px 0 rgba(0, 0, 0, 0.25);
 }
+
+// reset bulma pre styles
```
[the remark this answers is not in the document]
Comment (mplattner) [reply docs/src/assets/style.scss:99]: Sure, I forgot that after adding more rules.

File: docs/scripts/clirecording/clidirector.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+# todo: merge with CliDirector
+class MitmCliDirector(CliDirector):
+    def __init__(self):
+        super().__init__()
+        self.instructions: typing.List[InstructionSpec] = []
+
+    def instruction(self, instruction: str, duration: float = 3, time_from: typing.Optional[float] = None, correction: float = 0) -> None:
+        if time_from is None:
+            time_from = self.current_time
+        time_from_str = str(datetime.timedelta(seconds = int(time_from + correction)))[2:]
```
[the remark this answers is not in the document]
Comment (mplattner) [reply docs/scripts/clirecording/clidirector.py:146]: Great idea.

File: docs/scripts/clirecording/screenplays.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    pane_bottom = window.split_window(attach=True)
+    pane_bottom.resize_pane(height=12)
+
+    d.focus_pane(pane_bottom)
+    d.pause(2)
+
+    d.type("curl")
+    d.message("Use curl’s `-x` option to specify a proxy, e.g., `curl -x http://127.0.0.1:8080` to use mitmproxy.")
+
+    d.type(" -x http://127.0.0.1:8080")
```
[the remark this answers is not in the document]
Comment (mplattner) [reply docs/scripts/clirecording/screenplays.py:30]: Yep, will change that. 👍

File: docs/scripts/clirecording/screenplays.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+    d.message("mitmproxy is controlled using keyboard shortcuts.")
+    d.message("Use your arrow keys `↑` and `↓` to change the focused flow (`>>`).")
+    d.press_key("Down", count=3, pause=0.5)
+    d.press_key("Up", count=2, pause=0.5)
+    d.press_key("Down", count=2, pause=0.5)
+
+    d.message("The focused flow (`>>`) is used as a target for various commands.")
+
+    d.message("One such command shows the flow details, it is bound to `↵`.")
```
[the remark this answers is not in the document]
Comment (mplattner) [reply docs/scripts/clirecording/screenplays.py:68]: Hm. Both should be fine. We can also go with `ENTER`.

File: docs/scripts/clirecording/clidirector.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        if target is None:
+            target = self.tmux_pane
+        for i in range(count):
+            if keys == " ":
+                keys = "Space"
+            target.send_keys(cmd=keys, enter=False, suppress_history=False)
+            self.pause(pause + random.uniform(0, self.pause_between_keys_rand))
+
+    def type(self, keys: str, pause: typing.Optional[float] = None, target = None) -> None:
+        if pause is None:
```
[the remark this answers is not in the document]
Comment (mplattner) [reply docs/scripts/clirecording/clidirector.py:65]: Great idea. Future work though 😄

**Comment by mplattner** [comment]
An updated preview is available at https://plattner.me/mp-docs/
Happy to hear your feedback. :)

**Review by mhils — COMMENTED** [review]
Very cool! I'm super happy overall, the new typing looks _so_ much more human. 😃 

My only change request would be changing `/concepts-advancedfeatures/` back to `/overview-features/`. That URL may not be ideal, but that page is linked from a whole bunch of StackOverflow/GitHub/... posts. I think Hugo should allow us to just fix the id so that we don't break all that. :)

**Comment by mplattner** [comment]
Cool, thanks 😄 

So the features page should stay in the `Core concepts` menu, but the URL should be `/overview-features/`, correct? (Or move Features back to the `Overview` section?

**Comment by mhils** [comment]
> So the features page should stay in the Core concepts menu, but the URL should be /overview-features/, correct?

Yes, this. FWIW, I feel "Features" may still fit better than "Advanced Features", but no strong opinion on this. Overall we're achieving substantial improvements here. :)

**Comment by mplattner** [comment]
Ok 👍 
We have `Features` in the `Introduction`, listing very basic features. The features in `Advanced Features` seem to go beyond mitmproxy's core feature of just inspecting flows, also reflected by the fact that (I think) most features there are addons. The `advanced` list lacks the basic core features, so it might be misleading to simply call it features. (I guess many users find mitmproxy useful without using any of the `advanced` features.) This was my rationale. However, I have no strong opinion either. :)

**Comment by mhils** [comment]
> also reflected by the fact that (I think) most features there are addons. 

I guess we don't see "addons" in mitmproxy as advanced/optional - for example, interception itself is an addon, and the proxy server itself will be (hosted by) an addon soon. I'd also argue that modifybody etc. aren't super advanced. :) Anyways - this PR looks fantastic, let's get the web tests fixed and this is good to go! 😃

**Comment by mplattner** [comment]
Oops, good that we have those tests in place. 👍 Should be fixed now 🤞

**Comment by mhils** [comment]
Awesome, thanks @mplattner! That's a fantastic GSoC conclusion. 🍰 🚀 🎉

**Comment by mplattner** [comment]
Whoop whoop, awesome! 🎉 😃 Many thanks again!

**Comment by Prinzhorn** [comment]
Nice, including it directly in the docs was a big brain move :smile:

**Pull request merged**